crypto/evp/signature.c

```diff
@@ -615,12 +873,77 @@ static int evp_pkey_signature_init(EVP_PKEY_CTX *ctx, int operation,
 
 int EVP_PKEY_sign_init(EVP_PKEY_CTX *ctx)
 {
-    return evp_pkey_signature_init(ctx, EVP_PKEY_OP_SIGN, NULL);
+    return evp_pkey_signature_init(ctx, NULL, EVP_PKEY_OP_SIGN, NULL);
 }
 
 int EVP_PKEY_sign_init_ex(EVP_PKEY_CTX *ctx, const OSSL_PARAM params[])
 {
-    return evp_pkey_signature_init(ctx, EVP_PKEY_OP_SIGN, params);
+    return evp_pkey_signature_init(ctx, NULL, EVP_PKEY_OP_SIGN, params);
+}
+
+int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *ctx,
+                           EVP_SIGNATURE *algo, const OSSL_PARAM params[])
+{
+    if (algo == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+        return -1;
+    }
+    return evp_pkey_signature_init(ctx, algo, EVP_PKEY_OP_SIGN, params);
+}
+
+int EVP_PKEY_sign_message_init(EVP_PKEY_CTX *ctx,
+                               EVP_SIGNATURE *algo, const OSSL_PARAM params[])
+{
+    if (algo == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+        return -1;
+    }
```

Comment:
I disagree with these `algo == NULL` checks added as there is no reason why EVP_PKEY_sign_init_ex2() could not collapse to EVP_PKEY_sign_init_ex() when NULL `algo` is used. 

And the EVP_PKEY_sign_message_init() makes sense with `algo == NULL` for EdDSA - there you can look up the operation from the key type.

Reply:
Yup, I'll back off this